openscenegraph: support more file formats

aumuell opened the pull request:
add support for all formats for which the required libraries are provided by spack,
each additional dependency is optional

aumuell committed:
openscenegraph: compatibility with openexr@3

apply patch from gentoo

aumuell committed:
openscenegraph: add versions for git branches

OpenSceneGraph did not see a release in a long time, but the branch for the current stable release 3.6 has received many changes

aumuell committed:
openscenegraph: support more file formats

optionally read all formats for which spack has support libraries

aumuell committed:
openscenegraph: adopt package

aumuell commented:
@spackbot fix Style

aumuell committed:
openscenegraph: fix style

alalazo reviewed (commented):
Are all the variants besides `app` activated automatically during build?

aumuell commented:
> Are all the variants besides `app` activated automatically during build?

I'm not entirely sure if I understand correctly.
Do you mean that if e.g. `^gdal` is in openscenegraph's spec, it enables `+gdal`? Then the answer is no. If it should be done, is there an example of a package which does this?
Otherwise I did not get it.

alalazo commented on var/spack/repos/builtin/packages/openscenegraph/package.py:
```diff
@@ -59,8 +97,8 @@ def cmake_args(self):
             "-DDYNAMIC_OPENSCENEGRAPH={0}".format(shared_status),
             "-DDYNAMIC_OPENTHREADS={0}".format(shared_status),
             "-DOPENGL_PROFILE={0}".format(opengl_profile),
+            self.define_from_variant("BUILD_OSG_APPLICATIONS", "apps"),
```
What I mean is: you added 9 variants, and this is the only Cmake option that changed. Is that correct? Is CMake auto-detecting the other dependencies? If so, can we use explicit options to enable/disable them according to the variant?

aumuell committed:
openscenegraph: explicitly switch plugins according to variants

aumuell committed:
openscenegraph: typo

aumuell committed:
openscenegraph: remove special treatment for ffmpeg

disable building of ffmpeg plugin instead of setting ffmpeg CMake variables to useless values

aumuell committed:
openscenegraph: fix style

aumuell commented:
It would enable the plugins according to the dependencies being found. But of course this might lead to finding dependencies outside of spack's realm. Explicitly disabling the plugins will still let it find all the dependencies, but they won't be used.

And it will create CMake warnings because of unused settings from the command line, if the dependency for a plugin was not found, as it would not even try to enable it without the dependency. But this price is probably worth paying.

aumuell committed:
openscenegraph: ilmbase is not a dependency when using openexr@3

ilmbase is only required when linking against openexr@2, but it creates problems with
conflicting headers when used together with imath, which is a dependency of openexr@3.
Removing the dependency on ilmbase fixes this.

alalazo commented on var/spack/repos/builtin/packages/openscenegraph/package.py:
```diff
@@ -59,22 +97,35 @@ def cmake_args(self):
+                value = 0
+            return self.define("BUILD_OSG_PLUGIN_{}".format(plugin.upper()), value)
+
+        args.append(build_plugin("dicom", "dcmtk"))
+        args.append(build_plugin("exr", "openexr"))
+        args.append(build_plugin("ffmpeg"))
+        args.append(build_plugin("gdal"))
+        args.append(build_plugin("ogr", "gdal"))
+        args.append(build_plugin("gta"))
+        args.append(build_plugin("inventor"))
+        args.append(build_plugin("opencascade"))
+        args.append(build_plugin("pdf"))
+        args.append(build_plugin("svg"))
```
Apologies for getting back to this kind of late. Can you use `self.define_from_variants` instead of redefining a similar function in the package?

aumuell replied:
Thanks for this suggestion: while looking at the resulting diff, I saw that it totally makes sense not to reinvent this wheel!

aumuell committed:
openscenegraph: use self.define_from_variant

there was really not much to be gained from reimplementing something
similiar

aumuell committed:
openscenegraph: use self.define{,_from_variant}

alalazo reviewed (approved):
Thanks!
